apps/finance/app/src/components/useFilteredTransfers.js

```diff
@@ -0,0 +1,106 @@
+import { useState, useCallback, useMemo } from 'react'
+import { endOfDay, isWithinInterval, startOfDay } from 'date-fns'
+import { TRANSFER_TYPES, Incoming, Outgoing } from '../transfer-types'
+import { addressesEqual } from '../lib/web3-utils'
+
+const UNSELECTED_TOKEN_FILTER = -1
+const UNSELECTED_TRANSFER_TYPE_FILTER = -1
+const UNSELECTED_DATE_RANGE_FILTER = { start: null, end: null }
+
+function useFilteredTransfers({ transactions, tokens }) {
+  const [page, setPage] = useState(0)
+  const [selectedDateRange, setSelectedDateRange] = useState(
+    UNSELECTED_DATE_RANGE_FILTER
+  )
+  const [selectedTransferType, setSelectedTransferType] = useState(
+    UNSELECTED_TRANSFER_TYPE_FILTER
+  )
+  const [selectedToken, setSelectedToken] = useState(UNSELECTED_TOKEN_FILTER)
+  const handleSelectedDateRangeChange = useCallback(range => {
+    setPage(0)
+    setSelectedDateRange(range)
+  }, [])
+  const handleTokenChange = useCallback(index => {
+    setPage(0)
+    setSelectedToken(index || UNSELECTED_TOKEN_FILTER)
+  }, [])
+  const handleTransferTypeChange = useCallback(index => {
+    setPage(0)
+    setSelectedTransferType(index || UNSELECTED_TRANSFER_TYPE_FILTER)
+  }, [])
+  const handleClearFilters = useCallback(() => {
+    setPage(0)
+    setSelectedTransferType(UNSELECTED_TRANSFER_TYPE_FILTER)
+    setSelectedToken(UNSELECTED_TOKEN_FILTER)
+    setSelectedDateRange(UNSELECTED_DATE_RANGE_FILTER)
+  }, [])
+
+  const tokensToFilter = useMemo(() => [{ symbol: 'All tokens' }, ...tokens], [
+    tokens,
+  ])
+
+  const filteredTransfers = useMemo(
+    () =>
+      transactions.filter(({ date, isIncoming, token }) => {
+        const type = isIncoming ? Incoming : Outgoing
```

Comment:
I'm not feeling super good about this particular part; it seems fine for what it does, but maybe there is a better way to get the type, as for the transactions we only get an `isIncoming` boolean.

Reply:
Ah, where do we get `isIncoming`? Is it from the script? If so, we could use the type directly there or marshall it in the `app-state-reducer` so we avoid this.

Reply:
Marshalling it in the `app-state-reducer` sounds good 👍 , let's do this in another PR.